pkg/controller/nodeobservatory/controller.go

```diff
@@ -0,0 +1,379 @@
+				n.handlersMux.RLock()
+				defer n.handlersMux.RUnlock()
+				for _, updateHandler := range n.updateEventHandlers {
+					updateHandler(kluster, oldObj.(*api_v1.Node), newObj.(*api_v1.Node))
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if kluster, err = n.getKlusterByKey(key); err != nil {
+					n.logger.Log(err)
+					return
+				}
+				n.handlersMux.RLock()
+				defer n.handlersMux.RUnlock()
+				for _, deleteHandler := range n.deleteEventHandlers {
+					deleteHandler(kluster, obj.(*api_v1.Node))
+				}
+			},
+		})
+
+		n.nodeInformerMap.Store(
+			key,
+			nodeInformer,
+		)
+
+		cache.WaitForCacheSync(
+			n.stopCh,
+			nodeInformer.HasSynced,
+		)
+
+		go func(informer *NodeInformer) {
+			n.informerWg.Add(1)
+			defer n.informerWg.Done()
+			for {
+				select {
+				case <-n.stopCh:
+					informer.close()
+					return
+				default:
+					informer.run()
+				}
+			}
+		}(nodeInformer)
+
+	}
+	return nil
+}
+
+func (n *NodeObservatory) getKlusterByKey(key string) (*v1.Kluster, error) {
+	o, exists, err := n.Factories.Kubernikus.Kubernikus().V1().Klusters().Informer().GetIndexer().GetByKey(key)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("kluster %v was not found", key)
+	}
+	return o.(*v1.Kluster), err
+}
+
+func (n *NodeObservatory) klusterAddFunc(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		n.queue.Add(key)
+	}
+}
+
+func (n *NodeObservatory) klusterUpdateFunc(cur, old interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(cur)
+	if err == nil {
+		n.queue.Add(key)
+	}
+}
+
+func (n *NodeObservatory) klusterDeleteFunc(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err == nil {
+		n.queue.Add(key)
+	}
+}
+
+// GetStoreForKluster returns the SharedIndexInformers cache.Store for a given kluster or an error
+func (n *NodeObservatory) GetStoreForKluster(kluster *v1.Kluster) (cache.Store, error) {
+	informer, err := n.getNodeInformerForKluster(kluster)
+	if err != nil {
+		return nil, err
+	}
+	return informer.GetStore(), nil
+}
+
+// GetIndexerForKluster returns the SharedIndexInformers cache.Indexer for a given kluster or an error
+func (n *NodeObservatory) GetIndexerForKluster(kluster *v1.Kluster) (cache.Indexer, error) {
+	informer, err := n.getNodeInformerForKluster(kluster)
+	if err != nil {
+		return nil, err
+	}
+	return informer.GetIndexer(), nil
+}
+
+func (n *NodeObservatory) getNodeInformerForKluster(kluster *v1.Kluster) (NodeInformer, error) {
+	key, err := cache.MetaNamespaceKeyFunc(kluster)
+	if err != nil {
+		return NodeInformer{}, err
+	}
+	informer, ok := n.nodeInformerMap.Load(key)
+	if ok {
+		return informer.(NodeInformer), nil
+	}
+	return NodeInformer{}, fmt.Errorf("no informer found for kluster %v", key)
+}
+
+// HasSyncedForKluster returns true if the store of the kluster's SharedIndexInformer has synced.
+func (n *NodeObservatory) HasSyncedForKluster(kluster *v1.Kluster) bool {
```

Comment:
I'm unsure if we need this function...

Reply:
It's provided by the SharedIndexInformer. If you don't see a use-case we can just remove it.